Vue Spring Boot CRUD updated (#1244)
oktadeveloper/okta-blog: _source/_posts/2022-08-19-build-crud-spring-and-vue.md

This project has two major parts:

- Spring Boot API
- Vue client

The Spring Boot app will include an H2 in-memory database and will use Spring Data JPA to map our todo data model to a database table for persistence. As you'll see, the server will leverage Spring Boot's ability to quickly expose data via a REST API with a minimum of configuration.

The client will use [Vue 3](https://vuejs.org/) and the Quasar framework. [The Quasar framework](https://quasar.dev/) provides components and layout tools to help build Vue applications quickly with a consistent, high-quality user interface.

**Prerequisites:**

- [Java 11](https://adoptium.net/): or use [SDKMAN!](https://sdkman.io/) to manage and install multiple versions
- [Okta CLI](https://cli.okta.com/manual/#installation)
- [HTTPie](https://httpie.org/doc#installation)
- [Node 16+](https://nodejs.org)
- [Vue CLI](https://cli.vuejs.org/guide/installation.html)

You will need a free Okta Developer account if you don't already have one. But you can wait until later in the tutorial and use the Okta CLI to log in or register for a new account.

{% include toc.md %}

> 

Reviewer: I would mention that they can clone the repo and configure it to use Okta with the instructions in its README. That way, if they want to just get something working quickly, they can. 

I plan on replacing the current contents of https://github.com/oktadev/okta-spring-boot-vue-crud-example with this post. I'll do that in a PR that combines your current repo with the results from my QA.

## Create Okta OIDC app

Open a Bash shell. Create a parent directory for the project. Eventually, this will include both the resource server and client projects.

```bash
mkdir spring-boot-vue-crud
cd spring-boot-vue-crud
```

{% include setup/cli.md type="spa" framework="Vue" loginRedirectUri="http://localhost:8080/callback" %}

Your console should finish with something like this:

```bash
...
Configuring a new OIDC Application, almost done:
Created OIDC application, client-id: 0oa5rdbh01J0293u209
|
Okta application configuration: 
Issuer:    https://dev-123567.okta.com/oauth2/default
Client ID: 0oa5rdbh01J0293u209

> ```

Reviewer: If you look at the rendered preview of this post, you'll see this same information under the "What does the Okta CLI do?" expandable section. For that reason, you might want to remove this and just say something like "The client ID and issuer URI will be printed to your terminal. Copy them somewhere safe."

```bash
curl https://start.spring.io/starter.tgz \
  -d bootVersion=2.6.10 \
  -d javaVersion=11 \
  -d dependencies=web,data-rest,lombok,data-jpa,h2,okta \
  -d type=gradle-project \
  -d baseDir=resource-server \
| tar -xzvf - && cd resource-server
```

The dependencies you're including are:

- `web`: [Spring Web MVC](https://docs.spring.io/spring-framework/docs/3.2.x/spring-framework-reference/html/mvc.html), adds basic HTTP REST functionality
- `data-jpa`: [Spring Data JPA](https://spring.io/projects/spring-data-jpa), makes it easy to create JPA-based repositories
- `data-rest`: [Spring Data REST](https://spring.io/projects/spring-data-rest), exposes Spring Data repositories as resource servers
- `h2`: the [H2](https://www.h2database.com/html/main.html) in-memory database used for demonstration purposes
- `lombok`: [Project Lombok](https://projectlombok.org/), adds some helpful annotations that eliminate the need to write a lot of getters and setters
- `okta`: [Okta Spring Boot Starter](https://github.com/okta/okta-spring-boot) that helps OAuth 2.0 and OIDC configuration

> Project Lombok saves a lot of clutter and ceremony code. However, if you're using an IDE, **you'll need to install a plugin for Lombok**.

Reviewer: ```suggestion
Project Lombok saves a lot of clutter and ceremony code. However, if you're using an IDE, **you'll need to install a plugin for Lombok**. See the [project's installation docs](https://projectlombok.org/) for more information.
```

`src/main/resources/application.properties`

```properties
server.port=9000
okta.oauth2.issuer=<your-issuer-uri>
okta.oauth2.clientId=<your-client-id>
```

**You need to replace the two bracketed values** with the values you generated above for the OIDC app using the Okta CLI.

You can run the bootstrapped project right now and see if it starts. It should start but won't do much.

```bash
./gradlew bootRun
```

Create a `SecurityConfiguration` class to configure Spring Security. The class below configures web security to allow all requests, effectively bypassing security. This is just so you can test the resource server initially. You'll enable security shortly.

`src/main/java/com/example/demo/SecurityConfiguration.java`

```java
package com.example.demo;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.web.SecurityFilterChain;

@EnableWebSecurity

> @Configuration

Reviewer: You don't need `@EnableWebSecurity` here. `@Configuration` should be enough.

public class SecurityConfiguration {

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .anyRequest().permitAll();

>         return http.build();

Reviewer: My IntelliJ formatting is a bit different. 

```java
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http.authorizeRequests()
            .anyRequest().permitAll();
        return http.build();
    }
```

If you're using the defaults, it's fine to leave it as-is, but I think mine looks a little better.

Now, create the data model for the todo items. 

`src/main/java/com/example/demo/Todo.java`

```java
package com.example.demo;

import lombok.*;

import javax.persistence.Id;
import javax.persistence.GeneratedValue;
import javax.persistence.Entity;

@Entity
@Data
@NoArgsConstructor
public class Todo {

    @Id @GeneratedValue
    private Long id;

    @NonNull
    private String title;

    private Boolean completed = false;

}
```

Notice the use of the Lombok annotations (`@Entity`, `@Data`, and `@NoArgsConstructor`) to keep the code simple and clean.

The todo items have two fields: a title string and a completed boolean. The fields are annotated with Spring Data JPA annotations that allow the Java class to be mapped to a database table for persistence.

Create a repository to persist the data model.

`src/main/java/com/example/demo/TodoRepository.java`

```java
package com.example.demo;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.rest.core.annotation.RepositoryRestResource;

@RepositoryRestResource
interface TodoRepository extends JpaRepository<Todo, Long> {}
```

> This is a Spring JpaRepository that can persist the data model you just defined. Because it is annotated with `@RepositoryRestResource` (and because the `data-rest` dependency was included), this repository will be automatically exposed as a web resource.

Reviewer: ```suggestion
This is a Spring Data JpaRepository that can persist the data model you just defined. Because it is annotated with `@RepositoryRestResource` (and because the `data-rest` dependency was included), this repository will be automatically exposed as a web resource.
```

## Test the unsecured app and secure it

Run the app using the following command from the `resource-server` subdirectory.

```bash
./gradlew bootRun
```

Open a new Bash shell and use HTTPie to test the resource server.

```bash
http :9000/todos
```

> 

Reviewer: ```suggestion

You should see a response like the following:

```

```bash
HTTP/1.1 200 
...

{
    "_embedded": {
        "todos": [
            {
                "_links": {
                    "self": {
                        "href": "http://localhost:9000/todos/1"
                    },
                    "todo": {
                        "href": "http://localhost:9000/todos/1"
                    }
                },
                "completed": false,
                "id": 1,
                "title": "Buy milk"
            },
            {
                "_links": {
                    "self": {
                        "href": "http://localhost:9000/todos/2"
                    },
                    "todo": {
                        "href": "http://localhost:9000/todos/2"
                    }
                },
                "completed": true,
                "id": 2,
                "title": "Eat pizza"
            },
            ...
        ]
    },

> ...

Reviewer: ```suggestion
    ...
```

Reviewer: Update to align with the previous content.

}
```

> Stop the resource server using `control-c`. 

Reviewer: ```suggestion
Stop the resource server using `CTRL + C`. 
```

> Edit the security configuration file, updating the security configuration bean definition to the following.

Reviewer: ```suggestion
Edit the `SecurityConfiguration.java` file and change the filter chain's bean definition to enable a resource server.
```

`src/main/java/com/example/demo/OAuth2ResourceServerSecurityConfiguration.java`

```java
@Bean
public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
    http.authorizeRequests()
            .anyRequest().authenticated()
            .and()
            .oauth2ResourceServer().jwt();

>     return http.build();

Reviewer: My IntelliJ instances formats this as:

```java
        http.authorizeRequests()
            .anyRequest().authenticated()
            .and()
            .oauth2ResourceServer().jwt();
        return http.build();
```

}
```

This configuration requires JWT auth on all requests.

> Restart the server. Use `control-c` to stop it if it's running.

Reviewer: ```suggestion
Restart the server. Use `CTRL + C` to stop it if it's running.
```

```bash
./gradlew bootRun
```

Use HTTPie again to try and request the todo items.

```bash
http :9000/todos
```

You will get an error.

```bash
HTTP/1.1 401 
...

> 401 Unauthorized
> 

Reviewer: ```suggestion
401 Unauthorized
```

```

The resource server is finished. The next step is to create the Vue client.

## Create the Vue JavaScript client

> From the project parent directory, use the Vue CLI to create a new application and navigate into the newly created `client` directory.

Reviewer: ```suggestion
From the project's root directory, use the Vue CLI to create a new application and navigate into the newly created `client` directory. Install the Vue CLI if you don't have it installed with `npm i -g @vue/cli`.
```

```bash
vue create client
cd client
```

> Pick `Default Vue 3`

Reviewer: ```suggestion
Pick **Default ([Vue 3] babel, eslint)** when prompted.
```

Wait for it to finish. Add the Quasar framework.

```bash
vue add quasar
```

> You can just accept the defaults. For me they were the following.

Reviewer: ```suggestion
You can just accept the defaults. For me, they were the following.
```

> - **Allow Quasar to replace App.vue, About.vue, Home.vue and (if available) router.js?** Yes

Reviewer: I think the bolding in this list should be inverted. That is, the question should not be bold and the answers should be bold.

> - **Pick your favorite CSS preprocessor:** Sass

Reviewer: ```suggestion
- **Pick your favorite CSS preprocessor:** Sass with indented syntax
```

> - **Choose Quasar Icon Set:** Material

Reviewer: ```suggestion
- **Choose Quasar Icon Set:** Material Icons (recommended)
```

- **Default Quasar language pack:** en-US

> - **Use RTL support?** No

Reviewer: There's another option "Select features:", you can just press enter to select none.

> Add our dependencies.

Reviewer: ```suggestion
Add additional dependencies for HTTP requests, logging, routing, and authentication.
```

```bash

> npm i axios@^0.27.2 vuejs3-logger@^1.0.0 vue-router@^4.0.16 @okta/okta-vue@^5.3.0

Reviewer: We typically lock versions (with no ranges) since so many npm libraries break between minor releases. However, developers often ignore what we say and trim off the versions. You can do what you'd normally recommend here.

Reviewer: I get an error when I run this:

```
npm i axios@^0.27.2 vuejs3-logger@^1.0.0 vue-router@^4.0.16 @okta/okta-vue@^5.3.0
zsh: no matches found: axios@^0.27.2
```

This could be caused by zsh. I've seen issues like this before and quoting the string solves it. 

If I remove the ^, it works. 

```
npm i axios@0.27.2 vuejs3-logger@1.0.0 vue-router@4.0.16 @okta/okta-vue@5.3.0
```

```

- `axios` is an HTTP client request library
- `vuejs3-logger` is a logging library
- `vue-router` is the standard for routing between pages in Vue

> - `okta/okta-vue` is the Okta helper for Vue

Reviewer: I would make this list match the format that you used for the Spring Boot dependencies at the beginning. To abbreviate, use a colon and remove "is".

> To learn more about how Okta integrates with Vue, take a look at [the GitHub page](https://github.com/okta/okta-vue) for the `okta/okta-vue` project. There are also more resources and example applications listed on in [the general Okta docs](https://developer.okta.com/code/vue/).

Reviewer: ```suggestion
To learn more about how Okta integrates with Vue, take a look at [the GitHub page](https://github.com/okta/okta-vue) for the `okta/okta-vue` project. There are also more resources and example applications listed on in [the Okta docs for Vue](https://developer.okta.com/code/vue/).
```

Replace `main.js` with the following. Look at the `OktaAuth` configuration object. Notice the client ID and issuer URI are pulled from a `.env` file.

`src/main.js`

```js
import { createApp } from 'vue'
import App from './App.vue'
import {Quasar} from 'quasar'
import quasarUserOptions from './quasar-user-options'
import VueLogger from 'vuejs3-logger'
import router from './router'
import createApi from './Api'

import {OktaAuth} from '@okta/okta-auth-js'
import OktaVue from '@okta/okta-vue'

if (process.env.VUE_APP_ISSUER_URI == null || process.env.VUE_APP_CLIENT_ID == null || process.env.VUE_APP_SERVER_URI == null) {
  throw "Please define VUE_APP_ISSUER_URI, VUE_APP_CLIENT_ID, and VUE_APP_SERVER_URI in .env file"
}

const oktaAuth = new OktaAuth({

>   issuer: process.env.VUE_APP_ISSUER_URI,  // pulling from .env file

Reviewer: ```suggestion
  issuer: process.env.VUE_APP_ISSUER_URI,  // pulled from .env file
```

I think "pulled" sounds better than "pulling". "read" might work too.

  clientId: process.env.VUE_APP_CLIENT_ID,  // pulling from .env file
  redirectUri: window.location.origin + '/callback',
  scopes: ['openid', 'profile', 'email']
})

const options = {
  isEnabled: true,
  logLevel: 'debug',
  stringifyArguments: false,
  showLogLevel: true,
  showMethodName: false,
  separator: '|',
  showConsoleColors: true
};

const app = createApp(App)
  .use(Quasar, quasarUserOptions)
  .use(VueLogger, options)
  .use(OktaVue, {oktaAuth})
  .use(router)

app.config.globalProperties.$api = createApi(app.config.globalProperties.$auth)

app.mount('#app')
```

> Stated very briefly, the file above creates the main Vue app and configures it to use our dependencies: Quasar, VueLogger, OktaVue, and the router. It also creates the API class that handles the requests to the resource server and passes it the `$auth` object it needs to get the JWT.

Reviewer: ```suggestion
Stated very briefly, the file above creates the main Vue app and configures it to use the dependencies you added: Quasar, VueLogger, OktaVue, and the router. It also creates the API class that handles the requests to the resource server and passes it the `$auth` object it needs to get the JWT.
```

> Create a `.env` file in the client project root directory. The **Client ID** and **Issuer URI** are the values you used above in the Spring Boot `application.properties` file. The **Server URI** is the local URI for the Spring Boot API. You can leave this as it is unless you made a change (this gets used in the `Api.js` module).

Reviewer: ```suggestion
Create a `.env` file in the client project's root directory. The **Client ID** and **Issuer URI** are the values you used above in the Spring Boot `application.properties` file. The **Server URI** is the local URI for the Spring Boot API. You can leave this as it is unless you made a change (this gets used in the `Api.js` module).
```

Reviewer: Should module be changed to file?

`src/App.vue`

{% raw %}
```vue
<template>
  <q-layout view="hHh lpR fFf">

    <q-header elevated class="bg-primary text-white">
      <q-toolbar>
        <q-toolbar-title>
          <q-avatar>
            <q-icon name="kayaking" size="30px"></q-icon>
          </q-avatar>
          Todo App
        </q-toolbar-title>
        {{ this.claims && this.claims.email ? claims.email : '' }}
        <q-btn flat round dense icon="logout" v-if='authState && authState.isAuthenticated' @click="logout"/>
        <q-btn flat round dense icon="account_circle" v-else @click="login"/>
      </q-toolbar>
    </q-header>

    <q-page-container>
      <router-view></router-view>
    </q-page-container>

  </q-layout>
</template>

<script>
export default {
  name: 'LayoutDefault',
  data: function () {
    return {
      claims: null
    }
  },
  watch: {
    'authState.isAuthenticated'() {
      this.$log.debug(("watch triggered!"))
      this.updateClaims()
    }
  },
  created() {
    this.updateClaims()
  },
  methods: {
    async updateClaims() {
      if (this.authState && this.authState.isAuthenticated) {
        this.claims = await this.$auth.getUser()
      }
    },
    async login() {

>       await this.$auth.signInWithRedirect({originalUri: '/todos'})

Reviewer: ```suggestion
      await this.$auth.signInWithRedirect({ originalUri: '/todos' })
```

Create a new file to encapsulate the resource server access logic.

`src/Api.js`

```js
import axios from 'axios'

const instance = axios.create({
  baseURL: process.env.VUE_APP_SERVER_URI,
  timeout: 1000
});

const createApi = (auth) => {

  instance.interceptors.request.use(async function (config) {
    let accessToken = await auth.getAccessToken()
    config.headers = {
      Authorization: `Bearer ${accessToken}`
    }
    return config;
  }, function (error) {
    return Promise.reject(error);
  });

  return {

    // (C)reate
    createNew(text, completed) {
      return instance.post('/todos', {title: text, completed: completed})
    },

    // (R)ead
    getAll() {
      return instance.get('/todos', {
        transformResponse: [function (data) {
          return data ? JSON.parse(data)._embedded.todos : data;
        }]
      })
    },

    // (U)pdate
    updateForId(id, text, completed) {
      return instance.put('todos/' + id, {title: text, completed: completed})
    },

    // (D)elete
    removeForId(id) {
      return instance.delete('todos/' + id)
    }
  }
}

export default createApi
```

> All of the requests to the server go through this module. Notice that the server URI is hard coded at the top as `SERVER_URL`. In production this would be moved to a config file or an environment variable. Also take a look at how the access token is retrieved from the global `auth` object and injected into every request.

Reviewer: I'm not sure your commend about production makes sense. It's already in a config file and set as an environment variable, isn't it?

Create the router file.

`src/router/index.js`

```js

> import {createRouter, createWebHistory} from 'vue-router'
> import {navigationGuard} from '@okta/okta-vue'
> import Todos from "@/components/Todos";
> import Home from "@/components/Home";
> import {LoginCallback} from '@okta/okta-vue'

Reviewer: ```suggestion
import { createRouter, createWebHistory } from 'vue-router'
import { navigationGuard } from '@okta/okta-vue'
import Todos from "@/components/Todos";
import Home from "@/components/Home";
import { LoginCallback } from '@okta/okta-vue'
```

My IDE adds spaces around the brackets.

const routes = [
  {
    path: '/',
    component: Home
  },
  {
    path: '/todos',
    component: Todos,
    meta: {
      requiresAuth: true
    }
  },
  {path: '/callback', component: LoginCallback},
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
})

router.beforeEach(navigationGuard)

export default router
```

> The router has three paths. The home path and the todos path are straightforward. The last path, `/login/callback`, is provided by the Okta module to handle the login redirect from the Okta servers after authentication.

Reviewer: ```suggestion
The router has three paths. The home path and the todos path are straightforward. The last path, `/callback`, is provided by the Okta Vue SDK to handle the login redirect from the Okta servers after authentication.
```

  <div class="column justify-center items-center" id="row-container">
    <q-card class="my-card">
      <q-card-section style="text-align: center">
        <div v-if='authState && authState.isAuthenticated' >
          <h6 v-if="claims && claims.email">You are logged in as {{claims.email}}</h6>
          <h6 v-else>You are logged in</h6>
          <q-btn flat color="primary" @click="todo">Go to Todo app</q-btn>
          <q-btn flat @click="logout">Log out</q-btn>
        </div>
        <div v-else>
          <h6>Please <a href="#" @click.prevent="login">log in</a> to access Todo app</h6>
        </div>
      </q-card-section>
    </q-card>
  </div>
</template>

<script>
export default {
  name: "home-component",
  data: function () {
    return {
      claims: ''
    }
  },
  created () { this.setup() },
  methods: {
    async setup () {
      if (this.authState && this.authState.isAuthenticated) {
        this.claims = await this.$auth.getUser()
      }
    },
    todo() {
      this.$router.push("/todos")
    },
    async login () {
      await this.$auth.signInWithRedirect({ originalUri: '/todos'} )
    },
    async logout () {
      await this.$auth.signOut()
    }
  }
}

> </script>

Reviewer: My formatting for this is a bit different:

```vue
<template>
  <div class="column justify-center items-center" id="row-container">
    <q-card class="my-card">
      <q-card-section style="text-align: center">
        <div v-if='authState && authState.isAuthenticated'>
          <h6 v-if="claims && claims.email">You are logged in as {{ claims.email }}</h6>
          <h6 v-else>You are logged in</h6>
          <q-btn flat color="primary" @click="todo">Go to Todo app</q-btn>
          <q-btn flat @click="logout">Log out</q-btn>
        </div>
        <div v-else>
          <h6>Please <a href="#" @click.prevent="login">log in</a> to access Todo app</h6>
        </div>
      </q-card-section>
    </q-card>
  </div>
</template>

<script>
export default {
  name: "home-component",
  data: function () {
    return {
      claims: ''
    }
  },
  created() {
    this.setup()
  },
  methods: {
    async setup() {
      if (this.authState && this.authState.isAuthenticated) {
        this.claims = await this.$auth.getUser()
      }
    },
    todo() {
      this.$router.push("/todos")
    },
    async login() {
      await this.$auth.signInWithRedirect({ originalUri: '/todos' })
    },
    async logout() {
      await this.$auth.signOut()
    }
  }
}
</script>
```

```
{% endraw %}

Create the `TodoItem` component.

`src/components/TodoItem.vue`

{% raw %}
```vue
<template>
  <q-item-section avatar class="check-icon" v-if="this.item.completed">
    <q-icon color="green" name="done" @click="handleClickSetCompleted(false)"/>
  </q-item-section>
  <q-item-section avatar class="check-icon" v-else>
    <q-icon color="gray" name="check_box_outline_blank" @click="handleClickSetCompleted(true)"/>
  </q-item-section>

>   <q-item-section v-if="!editing">{{this.item.title}}</q-item-section>

Reviewer: Add spaces around brackets. I can do this once I'm done with my current QA.

      todo.title = value
    },

    handleShowError(message) {
      this.error = message
    },

    handleErrorClick () {
      this.error = null;
    },

  },

}
</script>

<style>
#row-container {
  margin-top: 100px;
}
.my-card {
  min-width: 600px;
}
.error {
  color: red;
  text-align: center;
  min-width: 600px;
  margin-top: 10px;
}
</style>
```
{% endraw %}

This component encapsulates the card that holds all of the todos, as well as the todo-associated interface elements. It also handles the rest of the functions related to updating todos on the server as well as in the local cache. 

You're welcome to delete the `HelloWorld.vue` component, if you want. Or you can leave it. It's not needed.

## Test the todo app

Make sure the Spring Boot API is still running. In a separate Bash shell, from the resource server directory, run the following command (if it is not already still running)

```bash
./gradlew bootRun
```

Start the Vue app using the embedded development server. From the client directory:

```bash
npm run serve
```

Open a browser and navigate to `http://localhost:8080`. You'll see the "please log in" page.

> {% img blog/spring-boot-crud-update/please-log-in.png alt:"Please log in" width:"1000" %}{: .center-image }

Reviewer: Mine looks slightly different on Chrome (104) and macOS Monterey (12.5.1).

<img width="657" alt="Screen Shot 2022-08-18 at 15 55 33" src="https://user-images.githubusercontent.com/17892/185502226-f67b9597-2c77-41cf-9d98-4d882adcaee6.png">

Log into the app using Okta's sign-in interface. 

{% img blog/spring-boot-crud-update/okta-login.png alt:"Okta SSO login" width:"600" %}{: .center-image }

That will redirect you to the Todo app's main screen.

{% img blog/spring-boot-crud-update/app-main-screen.png alt:"Todo app main screen" width:"1000" %}{: .center-image }

> 

Reviewer: I got an error when I first tried this. Refreshing the page solved the problem. I tried again in an incognito window and was unable to reproduce. ¯\_(ツ)_/¯

Reviewer: I was able to reproduce it again after updating vue-router. For this reason, we might want to increase the timeout or mention a refresh will fix. 

<img width="1790" alt="Screen Shot 2022-08-18 at 18 57 14" src="https://user-images.githubusercontent.com/17892/185519876-7da4d386-40e5-45b1-87aa-5bc73e56ddcf.png">

You should be able to delete items, add new items, rename, and filter items. All data is stored on the Spring Boot resource server and is presented by the Vue + Quasar frontend. 

> ## Wrapping up

Reviewer: ```suggestion
## Do More with Okta, Vue, and Spring Boot
```

Stolen from the post we're replacing.

> In this tutorial, you built a Spring Boot resource server backend and a Vue front-end. The Vue client used the latest Vue 3 version with the Quasar framework. The app included full CRUD (create, read, update, and delete) capabilities. It was all secured using Okta.

Reviewer: ```suggestion
In this tutorial, you built a Spring Boot resource server backend and a Vue front-end. The Vue client used the latest Vue 3 version with the Quasar framework. The app included full CRUD (create, read, update, and delete) capabilities. It was all secured using Okta.

You can find the source code for this example on GitHub, in the [@oktadev/okta-spring-boot-vue-crud-example](https://github.com/oktadev/okta-spring-boot-vue-crud-example) repository.

```

The client will use [Vue 3](https://vuejs.org/) and the Quasar framework. [The Quasar framework](https://quasar.dev/) provides components and layout tools to help build Vue applications quickly with a consistent, high-quality user interface.

Before you dig into the tutorial, I want to quickly introduce the technologies for those that might be unfamiliar. Feel free to skip down to the [prerequisites](#prerequisites) section if you're already familiar with Vue and Spring Boot.

{% include toc.md %}

## What is Vue.js?

Vue is a JavaScript view library, like React and Angular. It's designed to be incrementally adoptable, and the core library focuses solely on the view layer.

In my experience, Vue.js is a great alternative to React. I learned React first and came to use Vue later. Like React, Vue uses a virtual DOM, provides reactive and composable view components, and enforces a strict one-way parent-child relationship when defining properties and state. This means that it is performant and avoids many confusing state relationships that can occur without one-way data binding.

> However, unlike React, Vue uses templates instead of JSX (a potentially welcome and more immediately accessible option). Vue gives you component scoped CSS using style tags in single-file components. In practice, this difference is pretty significant because, in React, the JSX and CSS-like syntax are close enough to HTML and CSS to be confusing but not the same, which creates problems initially (ever gone from a language that doesn't require semicolons back to one that does? It's something like that).

Reviewer 2: ```suggestion
However, unlike React, Vue uses templates instead of JSX (a potentially welcome and more immediately accessible option). Vue gives you component-scoped CSS using style tags in single-file components. In practice, this difference is pretty significant because, in React, the JSX and CSS-like syntax are close enough to HTML and CSS to be confusing but not the same, which creates problems initially. (Ever gone from a language that doesn't require semicolons back to one that does? It's something like that.)
```

        this.newTodoTitle = ""
        this.todos.push({
          id: response.data.id,
          title: value,
          completed: false
        })
        this.$refs.newTodoInput.blur()
      }).catch((error) => {
        this.$log.debug(error);
        this.error = "Failed to add todo"
      });
    },
    handleCancelEditingNewTodo() {
      this.newTodoTitle = ""
    },

    handleSetCompleted(id, value) {
      let todo = this.todos.find(todo => id === todo.id)
      todo.completed = value
    },

    handleSetTitle(id, value) {
      let todo = this.todos.find(todo => id === todo.id)
      todo.title = value
    },

    handleShowError(message) {
      this.error = message
    },

    handleErrorClick() {
      this.error = null;
    },

  },

}
</script>

<style>
#row-container {
  margin-top: 100px;
}

.my-card {
  min-width: 600px;
}

.error {
  color: red;
  text-align: center;
  min-width: 600px;
  margin-top: 10px;
}
</style>
```
{% endraw %}

This component encapsulates the card that holds all of the todos and the todo-associated interface elements. It also handles the rest of the functions related to updating todos on the server and in the local cache.

> You're welcome to delete the `HelloWorld.vue` component, if you want. Or you can leave it. It's not needed.

Reviewer 2: ```suggestion
You're welcome to delete the `HelloWorld.vue` component if you want. Or you can leave it. It's not needed.
```

> ## Confirm your Spring Boot and Vue Todo app works

Reviewer 2: ```suggestion
## Confirm your Spring Boot and Vue todo app works
```

> Make sure the Spring Boot API is still running. In a separate Bash shell, from the resource server directory, run the following command (if it is not already still running)

Reviewer 2: ```suggestion
Make sure the Spring Boot API is still running. In a separate Bash shell, from the resource server directory, run the following command (if it is not already still running).
```
